File: lib/mappers/asset_mapper.dart
// lib/mappers/asset_mapper.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';  // ✅ debugPrint
import '../models/asset_model.dart';
import '../models/enums.dart';
import '../models/audit_history.dart';  // ✅ เพิ่ม
import '../configs/default_values.dart';

class AssetMapper {  static const Map<String, String> _assetClassMap = {
    'A2000': 'Building',
    'A2004': 'Machine',
    'A2007': 'Piping',
    'A2008': 'Plant Equipment',
    'A2009': 'Tools',
    'A2011': 'Furniture',
    'A3001': 'Vehicle',
    'A3002': 'Heavy Vehicle',
    'A5004': 'ROU Vehicle',
    'A62009': 'LVA Tools',
    'A62011': 'LVA Furniture',
    'A63001': 'LVA Vehicle',
    'A9000': 'LVA Equipment',
    'A9002': 'LVA Furniture',
    'A9004': 'LVA Vehicle',
    'A9801': 'AUC Machine',
    'A9802': 'AUC Other',
    'A2005': 'Machine (POM)',
  };

  static String getShortClassName(String rawClass) {
    return _assetClassMap[rawClass] ?? 'Unknown ($rawClass)';
  }

  static DateTime? parseTimestamp(dynamic value) {
    if (value == null) return null;
    if (value is DateTime) return value;
    if (value is Timestamp) return value.toDate();
    if (value is String) {
      try {
        return DateTime.parse(value);
      } catch (_) {
        return null;
      }
    }
    return null;
  }

  static AssetModel fromFirestore(Map<String, dynamic> json, String docId) {
    final assetNo = json['assetNo']?.toString() ?? docId;
    final description = json['description']?.toString() ?? DefaultValues.description;
    final assetClass = json['assetClass']?.toString() ?? '';
    final assetClassName = json['assetClassName']?.toString() ?? 
        getShortClassName(assetClass);
    final capDate = json['capDate']?.toString();
    final assetOwner = json['assetOwner']?.toString() ?? DefaultValues.owner;
    final costCenter = json['costCenter']?.toString() ?? DefaultValues.costCenter;
    final costCenterName = json['costCenterName']?.toString() ?? DefaultValues.costCenterName;
    final mainLocation = json['mainLocation']?.toString() ?? '';
    final lastLocationName = json['lastLocationName']?.toString() ?? '';
    
    final environment = Environment.fromString(
      json['environment']?.toString() ?? 'indoor'
    );
    final mobility = Mobility.fromString(
      json['mobility']?.toString() ?? 'fixed'
    );
    
    final status = json['status']?.toString() ?? '';
    final currentStatus = (json['currentStatus'] as num?)?.toInt() ?? 0;
    final lastImageUrl = json['lastImageUrl']?.toString() ?? '';
    final lastCondition = json['lastCondition']?.toString() ?? '';
    final remarks = json['remarks']?.toString();
    final updatedAt = parseTimestamp(json['updatedAt']);
    final updatedBy = json['updatedBy']?.toString() ?? '';

    // ✅ แปลง history จาก Map → List<AuditHistory>
    final List<AuditHistory> history = [];
    if (json['history'] is List) {
      final list = json['history'] as List;
      for (final item in list) {
        if (item is Map<String, dynamic>) {
          try {
            history.add(AuditHistory.fromJson(item));
          } catch (e) {
            debugPrint('⚠️ Failed to parse history item: $e');
          }
        }
      }
    }

    return AssetModel(
      assetNo: assetNo,
      description: description,
      assetClass: assetClass,
      assetClassName: assetClassName,
      capDate: capDate,
      assetOwner: assetOwner,
      costCenter: costCenter,
      costCenterName: costCenterName,
      mainLocation: mainLocation,
      lastLocationName: lastLocationName,
      environment: environment,
      mobility: mobility,
      status: status,
      currentStatus: currentStatus,
      lastImageUrl: lastImageUrl,
      lastCondition: lastCondition,
      remarks: remarks,
      updatedAt: updatedAt,
      updatedBy: updatedBy,
      history: history,  // ✅ เปลี่ยน
    );
  }

  static Map<String, dynamic> toJson(AssetModel asset) {
    return {
      'assetNo': asset.assetNo,
      'description': asset.description,
      'assetClass': asset.assetClass,
      'assetClassName': asset.assetClassName,
      'capDate': asset.capDate,
      'assetOwner': asset.assetOwner,
      'costCenter': asset.costCenter,
      'costCenterName': asset.costCenterName,
      'mainLocation': asset.mainLocation,
      'lastLocationName': asset.lastLocationName,
      'environment': asset.environment.toJson(),
      'mobility': asset.mobility.toJson(),
      'status': asset.status,
      'currentStatus': asset.currentStatus,
      'lastImageUrl': asset.lastImageUrl,
      'lastCondition': asset.lastCondition,
      'remarks': asset.remarks,
      'updatedAt': asset.updatedAt?.toIso8601String(),
      'updatedBy': asset.updatedBy,
      // ✅ แปลง history จาก List<AuditHistory> → List<Map>
      'history': asset.history.map((h) => h.toJson()).toList(),
    };
  }

  static AssetModel fromJson(Map<String, dynamic> json) {
    return fromFirestore(json, json['assetNo']?.toString() ?? '');
  }
}

File: AssetApp-Android/app/src/main/java/com/plbg/assetapp/data/mapper/AssetMapper.kt
package com.plbg.assetapp.data.mapper

import com.google.firebase.Timestamp
import com.plbg.assetapp.domain.model.AssetModel
import com.plbg.assetapp.domain.model.Environment
import com.plbg.assetapp.domain.model.Mobility
import com.plbg.assetapp.domain.model.DefaultValues
import java.time.LocalDateTime

object AssetMapper {
    private val assetClassMap = mapOf(
        "A2000" to "Building",
        "A2004" to "Machine",
        "A2007" to "Piping",
        "A2008" to "Plant Equipment",
        "A2009" to "Tools",
        "A2011" to "Furniture",
        "A3001" to "Vehicle",
        "A3002" to "Heavy Vehicle",
    )

    fun getShortClassName(rawClass: String): String =
        assetClassMap[rawClass] ?: "Unknown ($rawClass)"

    fun parseTimestamp(value: Any?): LocalDateTime? = when (value) {
        null -> null
        is LocalDateTime -> value
        is Timestamp -> value.toDate().toInstant()
            .atZone(java.time.ZoneId.systemDefault())
            .toLocalDateTime()
        is String -> runCatching { LocalDateTime.parse(value) }.getOrNull()
        else -> null
    }

    fun fromFirestore(json: Map<String, Any?>, docId: String): AssetModel {
        val assetNo = json["assetNo"]?.toString() ?: docId
        val assetClass = json["assetClass"]?.toString().orEmpty()

        val history = (json["history"] as? List<*>)
            ?.filterIsInstance<Map<String, Any?>>()
            ?.mapNotNull { runCatching { AuditHistoryMapper.fromJson(it) }.getOrNull() }
            ?: emptyList()

        return AssetModel(
            assetNo = assetNo,
            description = json["description"]?.toString() ?: DefaultValues.DESCRIPTION,
            assetClass = assetClass,
            assetClassName = json["assetClassName"]?.toString()
                ?: getShortClassName(assetClass),
            capDate = json["capDate"]?.toString(),
            assetOwner = json["assetOwner"]?.toString().orEmpty(),
            costCenter = json["costCenter"]?.toString().orEmpty(),
            costCenterName = json["costCenterName"]?.toString().orEmpty(),
            mainLocation = json["mainLocation"]?.toString().orEmpty(),
            lastLocationName = json["lastLocationName"]?.toString().orEmpty(),
            environment = Environment.fromString(json["environment"]?.toString()),
            mobility = Mobility.fromString(json["mobility"]?.toString()),
            status = json["status"]?.toString().orEmpty(),
            currentStatus = (json["currentStatus"] as? Number)?.toInt() ?: 0,
            lastImageUrl = json["lastImageUrl"]?.toString().orEmpty(),
            lastCondition = json["lastCondition"]?.toString().orEmpty(),
            remarks = json["remarks"]?.toString(),
            updatedAt = parseTimestamp(json["updatedAt"]),
            updatedBy = json["updatedBy"]?.toString().orEmpty(),
            history = history,
        )
    }

    fun toJson(asset: AssetModel): Map<String, Any?> = mapOf(
        "assetNo" to asset.assetNo,
        "description" to asset.description,
        "assetClass" to asset.assetClass,
        "assetClassName" to asset.assetClassName,
        "capDate" to asset.capDate,
        "assetOwner" to asset.assetOwner,
        "costCenter" to asset.costCenter,
        "costCenterName" to asset.costCenterName,
        "mainLocation" to asset.mainLocation,
        "lastLocationName" to asset.lastLocationName,
        "environment" to asset.environment.toJson(),
        "mobility" to asset.mobility.toJson(),
        "status" to asset.status,
        "currentStatus" to asset.currentStatus,
        "lastImageUrl" to asset.lastImageUrl,
        "lastCondition" to asset.lastCondition,
        "remarks" to asset.remarks,
        "updatedAt" to asset.updatedAt?.toString(),
        "updatedBy" to asset.updatedBy,
        "history" to asset.history.map { AuditHistoryMapper.toJson(it) },
    )
}